### DISCUSSION TITLE
Texture and Sampler Association

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#220] [Q&A] [answered]
Is it possible to associate texture samplers with specific textures? And I'm not sure if it matters, but is it still possible while using a global/shared pipeline resource signature (`CreatePipelineResourceSignature()`)?

At first, I set a single sampler with..
`GlobalSRB->GetVariableByName( ..., "texReg_sampler" )->Set( GlobalSampler );`

That worked great. No trouble. Now I'm trying to create a few samplers, and have textures select them based on certain traits, like being a 2D image vs 3D texture. I noticed in some of the samples, this is used..
`TextureView->SetSampler( SpecificSampler );`

But I don't understand how this associates anything with the shader SRB. When I try to do this without assigning anything to my global SRB, I get errors that nothing is assigned to it, and it renders like a default sampler.

Would you be able to point me in the right direction? Is there a way to synchronize textures and samplers? Do I need to do something on the HLSL side, or within the global signature? If I change my HLSL sampler into an array, will the texture sampler assignment then start working?

BTW, if you get some free time in the future, I would love to see more samples or tutorials that make use of shared pipeline resource signatures created with `CreatePipelineResourceSignature()`. Having only one sample (Tutorial 22) makes it challenging to understand some of the general concepts that are not used in that specific tutorial.

**Comment by TheMostDiligent** [comment, marked as answer]
Please read [this](https://github.com/DiligentGraphics/DiligentCore/blob/master/doc/TextureSamplers.md)

**Reply by RobertMtx** [reply]
Thanks, I had not seen that document. Is there an index of these somewhere? I also had no idea immutable samplers could be created with a shared resource signature. I am changing those over now. I will continue reading this. Thanks.

**Reply by RobertMtx** [reply]
I'm assuming the only way to select samplers based on texture traits of an array of textures is to use an array of samplers, similar to how tutorial 22 does it. Without an array, it would be difficult to select samplers per texture type, since multiple textures can be accessed in a single draw call.

I think I'll probably just change things around so that samplers are associated with pipeline states or renderers instead of individual textures. My frame rate is already low, so I don't want to add more heavy lifting to my drawing code. Thanks again for pointing me to the right document.

**Reply by TheMostDiligent** [reply]
Unless you care about OpenGL, which hopefully you don't, you should generally not use combined image samplers. Separate samplers are just like any other shader variable - you can control them like any other shader variable.

**Reply by RobertMtx** [reply]
I'm mainly targeting DX12, but trying not to do anything too specific for it, to make it easier to port later. But my main issue with control is that my textures are in an array, and each draw call can access any one of them, so it would be challenging to control which sampler is selected based on which texture is being rendered.

I mainly decided to associate the samplers with textures because it looked convenient at first. But now that I understand how it works internally, I can make a better design decision and hinge the sampler selection on what really matters - which is how the textures will be used, rather than what type of textures they are. I'll probably just define specific samplers for each situation, and let the shader code make use of whatever makes the most sense. Such as having the 2D drawing shaders use a clamped nearest point filter, and having most 3D shaders use anisotropic filtering. That kind of thing.

**Reply by TheMostDiligent** [reply]
Unless you are planning to support really ancient Android devices or Web, you should not care about OpenGL. In any case, OpenGL does not support dynamic texture indexing so what you are doing will likely not work on OpenGL anyway.

Samplers are significantly more scarce resource than textures. You can't have a large array of samplers - it will deplete the available sampler budget.
It is way more cleaner to separate samplers from texture and treat them indepenently.

**Reply by RobertMtx** [reply]
Yeah, as far as other APIs, I would only be concerned with newer APIs. By the time my game releases, DX13+Vulcan2 will likely be the standard, if that tells you anything. I've decided to go with immutable samplers. I only need a couple at the moment, but I think I read somewhere that their numbers are not limited like dynamic samplers.

**Reply by TheMostDiligent** [reply]
That is right, if you know what kind of sampler you need for your textures, which is a pretty often scenario, you should use immutable samplers. They are not limited and may be faster.
